src/meta.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * Managing `meta data` of each component that includes
+ * special method names and user defined decorator hooks. Since the property
+ * decorators are called before creating component constructor,
+ * we need to store decorator hooks in each prototype object and call it
+ * in component's class decorator later.
+ */
+
+import * as Vue from 'vue'
+import { Dictionary, VueInternal } from './declarations'
+
+export type DecoratorHook = (options: Vue.ComponentOptions<Vue>, value?: any) => void
+
+export class Meta {
+  /**
+   * The method names that the component decorator will
+   * just register on component options object.
+   * We can extend this list by using `Component.registerHooks`.
+   */
+  static internalHooks = [
+    'data', 'beforeCreate', 'created', 'beforeMount', 'mounted',
+    'beforeDestroy', 'destroyed', 'beforeUpdate', 'updated',
+    'activated', 'deactivated', 'render'
+  ]
+
+  /**
+   * Property, method and parameter decorators created by `createDecorator` helper
+   * will register functions that update component options for lazy processing.
+   * They will be executed just before creating component constructor.
+   */
+  decoratorMap: Dictionary<DecoratorHook> = {}
+
+  /**
+   * Property names that indicates the initial property will be
+   * processed on user-defined decorators. Also they will not collected in `data` hook.
+   */
+  decoratorAwaredKeys: Dictionary<boolean> = {}
+
+  get shouldGetInitalProperty () {
+    return Object.keys(this.decoratorAwaredKeys).length > 0
+  }
+}
+
+export interface CreateDecoratorOptions {
+  getInitialProperty?: boolean
+}
+
+// Property or method decorators
+export function createDecorator (
```

Comment:
TypeScript matches overriding call from top signature to bottom one.

That means, when we call `createDecorator`, the callback function is tested against the `factory` type in first overloading signature, which is `(opt, key, value: any)` here.  If two types match, the first overloading is picked up and the second overloading is not considered.

And the first `factory` type is the super type of the second one, so overloading will always match property decorators.